core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -297,25 +297,50 @@ impl ZkSyncStateKeeper {
             .with_context(|| format!("failed loading upgrade transaction for {protocol_version:?}"))
     }
 
-    async fn wait_for_new_batch_params(&mut self) -> Result<(SystemEnv, L1BatchEnv), Error> {
+    async fn wait_for_new_batch_params(
+        &mut self,
+        cursor: &IoCursor,
+    ) -> Result<(SystemEnv, L1BatchEnv), Error> {
         while !self.is_canceled() {
             if let Some(params) = self
                 .io
-                .wait_for_new_batch_params(POLL_WAIT_DURATION)
+                .wait_for_new_batch_params(cursor, POLL_WAIT_DURATION)
                 .await
                 .context("error waiting for new L1 batch params")?
             {
-                return Ok(params);
+                return Ok(self.map_batch_params(params, cursor).await?);
             }
         }
         Err(Error::Canceled)
     }
 
-    async fn wait_for_new_miniblock_params(&mut self) -> Result<MiniblockParams, Error> {
+    async fn map_batch_params(
+        &mut self,
+        params: L1BatchParams,
+        cursor: &IoCursor,
+    ) -> anyhow::Result<(SystemEnv, L1BatchEnv)> {
+        let contracts = self
+            .io
+            .load_base_system_contracts(params.protocol_version, cursor)
+            .await
+            .with_context(|| {
+                format!(
+                    "failed loading system contracts for protocol version {:?}",
+                    params.protocol_version
+                )
+            })?;
+        Ok(params.into_env(self.io.chain_id(), contracts, cursor))
+    }
```

Comment:
This method feels a bit out of place for the state keeper: we only use `IoCursor` (which is passed to IO in other methods already) and `self.io` here.
Moreover, `wait_for_new_batch_params` already receives `cursor` as an argument.

Maybe it's better to implement this logic as a part of IO trait?